### ISSUE TITLE
Constant high CPU in frontend

### ISSUE THREAD

**Issue by toupeira** [RocketChat/Rocket.Chat#5726]
Rocket.Chat Version: 0.49.3
Running Instances: 1
DB Replicaset OpLog: Enabled
Node Version: v4.7.2

Linux version: Debian sid
Firefox version: 50.1.0-1
Chromium version: 55.0.2883.75-5

Certain conversations in Rocket.Chat always hog my CPU (~30% on a quad-core system, distributed across all cores) and cause my fans to start spinning, this can be observed in Firefox using the Performance tab in the Devtools which shows a constant stream of "Recalculate Style" events. After removing one `<style>` / `<link>` tag after the other I think I've narrowed it down to the very first `<style>` tag in the `<head>` which has some rules for the `.loading-animation`. Either removing that `<style>` tag or running `$('.loading-animation').remove()` in the console instantly makes my CPU idle again.

Not sure why this only happens on certain conversations, I couldn't find any commonalities (I think I've ruled out GIFs, avatars, syntax highlighting, images, file uploads). But since I assume the loading animation is only needed during the initial load, I hope the tag can simply be removed after that's done?

**Comment by engelgabriel** [comment]
Thanks a lot @toupeira, this is very useful information. We do have an animation between changing channels, so maybe it is not being removed properly after the page load.

**Comment by engelgabriel** [comment]
@karlprieb I have just realised we don't remove the animation for `load more msgs`, we just push it upwards each time. Maybe we should remove the `$('.loading-animation')` and put it back when needed?

**Comment by karlprieb** [comment]
@engelgabriel Probably will decrease the CPU usage in this view. I will test that :)

**Comment by Abraka** [comment]
Whats about "low client cpus usage mode"? Creating settings in server administration...something like "Disabling all animation, gifs, zoutube videos etc" for slow or old clients PC?

**Comment by toupeira** [comment]
@Abraka since this depends on the client's CPU I think the setting should be per-client as well, and you can already disable most of these individually in the account settings.

**Comment by engelgabriel** [comment]
I agree, we should have a per user setting for that, disable all animations. We already have the function for that, used on our testing mode.

**Comment by karlprieb** [comment]
I removed all unnecessary "load more animations" to decrease CPU usage, but we need to analyze/profile the application to understand why our app increase CPU usage in some views.

After that we can create different issues to fix these problems.

**Comment by gdelavald** [comment]
Hey,
I've been looking into some cases where the load times and performance started to feel like a problem to the user.
I ran these analysis using a MacBook Pro with i5@2.7Ghz on OSx El Capitan with Chrome DevTools:

When trying to filter the emoji list I found some delays (also #6636):
<img width="1440" alt="timeline - filter emoji list" src="https://cloud.githubusercontent.com/assets/6303966/24825628/89c33f44-1bf8-11e7-8018-42cbe407bb48.png">

<img width="1198" alt="profiling - emoji picker" src="https://cloud.githubusercontent.com/assets/6303966/24825631/89c8fbb4-1bf8-11e7-9f4f-66e2fd64665d.png">

I did other tests using the case where a user would be switching to another channel.
First one has very few messages, it took like half a second to load the channel layout and a couple of messages:
<img width="1440" alt="timeline - switching to empty room" src="https://cloud.githubusercontent.com/assets/6303966/24825634/89e29bf0-1bf8-11e7-800b-14067ceff7f8.png">

Switching to a channel with a moderate amount of messages, from the click on the channel name to the finished loading took like 3 seconds (with 1.6s of these being only on one computation and almost all of it in the getRoomDom function) 
<img width="1440" alt="timeline - switching to room with history 2" src="https://cloud.githubusercontent.com/assets/6303966/24825629/89c459e2-1bf8-11e7-964f-2bf36738cf90.png">

Same as above, but with a few more messages and more loading time to render it.
<img width="1440" alt="timeline - switching to room with history" src="https://cloud.githubusercontent.com/assets/6303966/24825633/89ce609a-1bf8-11e7-8f18-54238cf18cd0.png">

Below there is one from the actual demo.rocket.chat, where the main computation takes 3.6s to render the general channel.
<img width="1031" alt="timeline - switching to general channel" src="https://cloud.githubusercontent.com/assets/6303966/24825635/89e598c8-1bf8-11e7-9555-11d6b7c38789.png">

I don't feel like this analysis covers all the app, but this cases already show some issues the application is having when rendering common views. One of the problems here is that this rendering times are not due to bad implementation but caused by the technology stack we have here. Perhaps going back to #1680 and looking into some new Library/Framework for the Views could be beneficial in the next steps.

@karlprieb let me know if (and how) you feel like this could be improved.

**Comment by ulope** [comment]
I'm also experiencing (what seems like) excessive CPU usage of constantly ~15% on a MacBook 2016 while RC is idle.

In my case I'm using the Electron App v. 2.10.5 with Server v. 0.64.1.

Using the internal dev tools I can see 'Recalculating Styles' (with the accompanying Compositing and Paint) events happening approximately every 16ms without any obvious cause.

The `.loading-animation` element mentioned above seems no longer to be there. Also, there are no animated avatars or reaction icons visible (although there are some of those used in our instance).

**Comment by toupeira** [comment]
@ulope I can confirm this with Chromium 66 after upgrading to Rocket.Chat 0.64.1, the paint events weren't happening previously with 0.62.1

**Comment by tassoevan** [comment]
Does anyone in this thread notice the same behavior after #12677 was merged (version 0.72.0)?

**Comment by seiichiro0185** [comment]
I'm still seeing a permanent CPU-Load of ~15-20% from Rocket.chat on Version 0.74.3

**Comment by tassoevan** [comment]
@seiichiro0185 May I ask you to record a performance profile on Firefox?

![peek 20-02-2019 11-39](https://user-images.githubusercontent.com/2263066/53099373-6dcfff80-3504-11e9-9b58-c6a06c8f1bf1.gif)

**Comment by seiichiro0185** [comment]
I did some more investigation. Seems like it is in part dependant on the Content. The High CPU-Usage of > 15% seems at least in Part due to Some animated Emojis that are used in our Rocketchat Instance. If no animated Gifs are displayed on screen I see a lower CPU-Usage of around 5% (which is still quite a bit in my Opinion for showing Website without any changes).

I created 2 Performance Profiles, one with the Gifs in view (rocketchat_gif.txt), and one without (rocketchat.txt). I had to rename .json to .txt, otherwise Github wouldn't let me upload them.

[rocketchat.txt](https://github.com/RocketChat/Rocket.Chat/files/2888445/rocketchat.txt)
[rocketchat_gif.txt](https://github.com/RocketChat/Rocket.Chat/files/2888448/rocketchat_gif.txt)

**Comment by catdad** [comment]
I was hoping there would be an update here. I am seeing issues of high CPU usage and the web app freezing for a good minute or two pretty regularly. I reported this on https://github.com/RocketChat/Rocket.Chat/issues/10900 but it looks like this is a similar report that has had more active history. When I look at Chrome's Task Manager, I see this:

![image](https://user-images.githubusercontent.com/2205537/73366985-cf542e80-427c-11ea-81f4-e831c92ff73d.png)

It seem to happens when I switch rooms or when I start typing into a room right after switching. I tried to debug -- though it's a bit difficult to do so when it happens at random -- but I did catch it in the profiler once and saw this:

![image](https://user-images.githubusercontent.com/2205537/73367107-ff9bcd00-427c-11ea-8ee5-28f742080731.png)

Note that it did not only hang for 9 seconds, but rather I only recorded 9 seconds so that the profiler doesn't crash. It hung for about 60 seconds or so total. When I clicked on the method where it hangs, it took me here:

![image](https://user-images.githubusercontent.com/2205537/73367185-2a862100-427d-11ea-9551-9eedd60d2bf2.png)

I hope this helps, as this is an issue that is preventing me from effectively communicating with my team. Sometimes, it's good to take a pause before answering a question, but most of the time, you want to have a quick and free-flowing conversation and it is hard to do if your chat app hangs for minutes at a time.
